src/bonsai/forms/spot.ts

```diff
@@ -0,0 +1,293 @@
+import { LAMPORTS_PER_SOL } from '@solana/web3.js';
+
+import { SpotApiCreateTransactionRequest, SpotApiSide, SpotApiTradeRoute } from '@/clients/spotApi';
+import { calc, mapIfPresent } from '@/lib/do';
+import { AttemptNumber } from '@/lib/numbers';
+
+import { createForm, createVanillaReducer } from '../lib/forms';
+import { ErrorType, simpleValidationError, ValidationError } from '../lib/validationErrors';
+
+export enum SpotSide {
+  BUY = 'BUY',
+  SELL = 'SELL',
+}
+
+export enum SpotBuyInputType {
+  USD = 'USD',
+  SOL = 'SOL',
+}
+
+export enum SpotSellInputType {
+  PERCENT = 'PERCENT',
+  SOL = 'SOL',
+}
+
+export interface SpotFormState {
+  side: SpotSide;
+  buyInputType: SpotBuyInputType;
+  sellInputType: SpotSellInputType;
+  size: string;
+}
+
+const initialState: SpotFormState = {
+  side: SpotSide.BUY,
+  buyInputType: SpotBuyInputType.USD,
+  sellInputType: SpotSellInputType.PERCENT,
+  size: '',
+};
+
+const reducer = createVanillaReducer({
+  initialState,
+  actions: {
+    setSide: (state, side: SpotSide) => ({
+      ...state,
+      side,
+      size: '', // Clear size when switching side
+    }),
+    setBuyInputType: (state, buyInputType: SpotBuyInputType) => ({
+      ...state,
+      buyInputType,
+    }),
+    setSellInputType: (state, sellInputType: SpotSellInputType) => ({
+      ...state,
+      sellInputType,
+    }),
+    setSize: (state, size: string) => ({
+      ...state,
+      size,
+    }),
+    reset: () => initialState,
+  },
+});
+
+export interface SpotFormInputData {
+  tokenPriceUsd: number | undefined;
+  solPriceUsd: number | undefined;
+  userSolBalance: number | undefined;
+  userTokenBalance: number | undefined;
+  tokenMint: string | undefined;
+  decimals: number | undefined;
+  pairAddress: string | undefined;
+  tradeRoute: SpotApiTradeRoute | undefined;
+  solanaAddress: string | undefined;
+}
+
+export interface SpotAmounts {
+  sol: number;
+  token: number;
+  usd: number;
+  percent: number | undefined;
+}
+
+export interface SpotSummaryData {
+  amounts: SpotAmounts | undefined;
+  payload: SpotApiCreateTransactionRequest | undefined;
+}
+
+function calculateSummary(state: SpotFormState, inputData: SpotFormInputData): SpotSummaryData {
+  const parsedSize = AttemptNumber(state.size);
+
+  const amounts: SpotAmounts | undefined = calc(() =>
+    mapIfPresent(
+      parsedSize,
+      inputData.tokenPriceUsd,
+      inputData.solPriceUsd,
+      (size, tokenUsdPrice, solUsdPrice): SpotAmounts => {
+        if (state.side === SpotSide.BUY) {
+          if (state.buyInputType === SpotBuyInputType.SOL) {
+            const sol = size;
+            const usd = sol * solUsdPrice;
+            const token = usd / tokenUsdPrice;
```

Comment:
i know we are using mapIfPresent, but should we also check for div 0 throughout the calculations?